947 - Modify csv export on the historic site search results

JessicaInTime opened the pull request:
Re: bcgov/BCHeritage#947

JrodS committed:
947 - Added query for csv export on heritage site

Refactored the databc query to include all cols for databc and the csv export in a single view

Both 0002 and 0004 files need to be ran against the db beforehand

bferguso reviewed (changes requested):
@jrods - Would you please implement this as a (or multiple) migration(s), complete with the reverse migration? The SQL files in the `pkg/` directory only get run when a project is recreated, so we can't use it to patch an existing project.

bferguso committed:
Merge pull request #40 from bcgov/bcrhp/public_website_docker

bcrhp/public website docker

JrodS committed:
947 - Added migration module for all the statements required for databc history one site and heritage site csv export

JrodS committed:
Merge remote-tracking branch 'origin/bcrhp/release/public_site' into 947_Modify-search-csv-export-historic-site

JrodS committed:
947 - Fix missing reference for csv export filtering on collection_ids (site_id col in the csv export view)

JrodS committed:
1.1.0 - Added release notes page for this page

bferguso commented on src/bcrhp/bcrhp/pkg/business_data/resource_views/0002_create_onerow_view.sql:
```diff
@@ -143,13 +143,35 @@ create materialized view mv_heritage_function as
                                                   __arches_get_concept_label(unnest(functional_state));
```
Please roll back any changes or additions made in the `pkg/business_data/resource_views` directory - these will never be run so all updates should happen in the migrations directory.

bferguso commented on src/bcrhp/bcrhp/pkg/business_data/resource_views/0004_historic_public_site_view.sql:
```diff
@@ -0,0 +1,219 @@
+drop view if exists databc.V_HISTORIC_ENVIRO_ONEROW_SITE;
```
Please roll back any changes or additions made in the `pkg/business_data/resource_views` directory - these will never be run so all updates should happen in the migrations directory.

bferguso commented on src/bcrhp/bcrhp/search/search_export.py:
```diff
@@ -0,0 +1,23 @@
+                instance_results = [dict(zip(instance_columns, row)) for row in cur.fetchall()]
+                print(instance_results[0])
+
+                return super().to_csv(instance_results, instance_columns, name)
```
I'm getting an `instance_columns` is not defined message here. Should this be `columns`?

bferguso commented on src/bcrhp/bcrhp/search/search_export.py:
```diff
@@ -0,0 +1,23 @@
+                """, [collection_ids])
+                columns = [desc[0] for desc in cur.description]
+                instance_results = [dict(zip(instance_columns, row)) for row in cur.fetchall()]
+                print(instance_results[0])
```
Please remove debug print statements

JessicaInTime replied:
oops, my bad, leftovers from a previous stash, fixed

bferguso commented on src/bcrhp/bcrhp/views/search.py:
```diff
@@ -0,0 +1,37 @@
+
+@group_required("Resource Exporter")
+def export_results(request):
+    # print("In BCGOv specific search results")
```
This is probably leftover from my dirty code, but please remove the debug statements

bferguso reviewed (changes requested):
Did all the existing materialized views need to be dropped and recreated? I looks like at least some of drop/recreate statements are redundant (eg mv_borden_number - not sure how many others)- we should only necessary changes in the migration and not touch the objects that aren't modified. This will keep QA and changes more localized and easier to digest.

When first running this w/ 1500 records it took quite a while to run. Once the SQL is trimmed down to what is actually changing we can have a look to see where we can improve performance.

JrodS committed:
947 - Reverted .sql in bcrhp/pkg/business_data/resource_views/, deprecated location and method for db changes

JrodS committed:
947 - Fixed bad logic for bcrhp search_export result fetching

JrodS committed:
947 - Removed unnecessary comments from bcrhp export_results

JessicaInTime commented:
> Did all the existing materialized views need to be dropped and recreated? I looks like at least some of drop/recreate statements are redundant (eg mv_borden_number - not sure how many others)- we should only necessary changes in the migration and not touch the objects that aren't modified. This will keep QA and changes more localized and easier to digest.

I wasn't sure what is within the deployed databases. If everything that is in the resource_views/0002_create_onerow_view.sql is in fact on the deploys, I can cut out the statements that are not modified in the mig module.

> When first running this w/ 1500 records it took quite a while to run. Once the SQL is trimmed down to what is actually changing we can have a look to see where we can improve performance.

The original databc view before modifying was also about the same amount of run time, if I remembered correctly. I haven't done any analyzing of the hotspots yet but will have a look

JrodS committed:
947 - Removed existing materialized views and functions for the databc and csv export

JrodS committed:
947 - Updated v_histori_site view to remove the get_first_address function usage in it's place with a subquery for improving performance

JrodS committed:
947 - Changed migration module name to have the ticket number as the prefix, same as arches core

bferguso commented on src/bcrhp/bcrhp/migrations/947_historic_site_one_line.py:
```diff
@@ -0,0 +1,356 @@
+            select
+                resourceinstanceid as site_id
+                , borden_number as "Borden Number"
+                , street_address as "Street Address"
```
For rows that can have a null value, please replace null value with empty string. Null values come out as "None" in the csv.

bferguso commented on src/bcrhp/bcrhp/migrations/947_historic_site_one_line.py:
```diff
@@ -0,0 +1,356 @@
+                , postal_code as "Postal Code"
+                , location_description as "Location Description"
+                , registration_status as "Registration Status"
+                , authorities->>'registry_types' as "Registry Type"
```
Please make array values a semi-colon separated string.

bferguso reviewed (changes requested):
@jrods - looking good so far! 
Please sort the export view by borden number and address the two changes requested below and then I think we're good to go. I'd like to do a a test of the DataBC view when we deliver to DLVR to ensure that the data looks the same.

JrodS committed:
947 - Addressed PR feedback for null and list values for the bcrhp csv export

bferguso commented on src/bcrhp/bcrhp/migrations/947_historic_site_one_line.py:
```diff
@@ -0,0 +1,368 @@
+            $$ language plpgsql;
+            """,
+            """
+            drop procedure if exists refresh_materialized_views;
```
This procedure existed before to refresh the DataBC view. This should reinstate the previous version rather than dropping it.

bferguso commented on src/bcrhp/bcrhp/migrations/947_historic_site_one_line.py:
```diff
@@ -0,0 +1,368 @@
+            ;
+            """,
+            """
+            drop view if exists databc.V_HISTORIC_ENVIRO_ONEROW_SITE cascade;
```
This needs to reinstate the previous version of this view.

JrodS committed:
947 - Changed backward migration for historic site

Need to keep the procedure as defined previously for refresh_materizlized_views and the databc.V_HISTORIC_ENVIRO_ONEROW_SITE

bferguso committed:
947 - Fix incorrect construction criteria, address logic, actor notes logic, incorrect drop of MV on the reverse migration

bferguso committed:
947 - Fix order of forward and reverse migrations

bferguso committed:
947 - Tweak chronology format, fix heritage class format, add missing name fields, order by BN

bferguso committed:
947 - Clean HTML from location description field
